squidpy/gr/_ripley.py

```diff
@@ -0,0 +1,245 @@
+
+    and then we apply a variance-stabilizying transformation:
+
+    .. math::
+
+        L(t) = (\frac{{K(t)}}{{\pi}})^{{1/2}}
+
+    For reference, check out
+    `Wikipedia <https://en.wikipedia.org/wiki/Spatial_descriptive_statistics#Ripley's_K_and_L_functions>`_
+    or :cite:`Baddeley2015-lm`.
+
+    Parameters
+    ----------
+    %(adata)s
+    %(cluster_key)s
+    mode
+        Which Ripley's statistic to compute.
+    %(spatial_key)s
+    metric
+        Which metric to use for computing distances.
+    n_neigh
+        Number of neighborhoods to consider for the knn graph.
+    n_simulations
+        How many simulations to run for computing pvalues.
+    n_observations
+        How many observations to generate for the spatial poisson point process.
+    max_dist
+        Maximum distances for the support. If `None`, `max_dist=`:math:`\sqrt{{area \over 2}}`.
+    n_steps
+        Number of steps for the support.
+    %(seed)s
+    %(copy)s
+
+    Returns
+    -------
+    %(ripley_stat_returns)s
+    """
+    _assert_categorical_obs(adata, key=cluster_key)
+    _assert_spatial_basis(adata, key=spatial_key)
+    coordinates = adata.obsm[spatial_key]
+    clusters = adata.obs[cluster_key].values
+
+    mode = RipleyStat(mode)  # type: ignore[assignment]
+    if TYPE_CHECKING:
+        assert isinstance(mode, RipleyStat)
+    mode = mode.s
+
+    # prepare support
+    N = coordinates.shape[0]
+    hull = ConvexHull(coordinates)
+    area = hull.volume
+    if max_dist is None:
+        max_dist = (area / 2) ** 0.5
+    support = np.linspace(0, max_dist, n_steps)
+
+    # prepare labels
+    le = LabelEncoder().fit(clusters)
+    cluster_idx = le.transform(clusters)
+    obs_arr = np.empty((le.classes_.shape[0], n_steps))
+
+    start = logg.info(
+        f"Calculating Ripley's {mode} statistic for `{le.classes_.shape[0]}` clusters and `{n_simulations}` simulations"
+    )
+
+    for i in np.arange(np.max(cluster_idx) + 1):
+        coord_c = coordinates[cluster_idx == i, :]
+        if mode == "F":
+            random = _ppp(hull, 1, n_observations, seed=seed)
+            tree_c = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(coord_c)
+            distances, _ = tree_c.kneighbors(random, n_neighbors=n_neigh)
+            bins, obs_stats = _f_g_function(distances.squeeze(), support)
+        elif mode == "G":
+            tree_c = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(coord_c)
+            distances, _ = tree_c.kneighbors(coordinates[cluster_idx != i, :], n_neighbors=n_neigh)
+            bins, obs_stats = _f_g_function(distances.squeeze(), support)
+        elif mode == "L":
+            distances = pdist(coord_c, metric=metric)
+            bins, obs_stats = _l_function(distances, support, N, area)
+        obs_arr[i] = obs_stats
+
+    sims = np.empty((n_simulations, len(bins)))
+    pvalues = np.ones((le.classes_.shape[0], len(bins)))
+
+    for i in range(n_simulations):
```

Comment:
Prime target for `_parallelize`. I know you've said optimize later and if you want, I can just add it once the other comments are addressed (shouldn't take long).

Reply:
yes but I feel that there will be a re-write of this in pure numba at some point (was contacted by someone who was interested in doing this). Since for the practical data at hand is not super needed right now (as it takes few seconds anyway, I think we can postpone it. And again like you said, would be straightforwad to add. Thanks for suggesting!

Reply:
Ok, agreed. After this is merged, could you please create an issue for this (maybe mark it as low-prio) so that it isn't forgotten?